src/main/kotlin/org/rust/lang/core/psi/impl/RustFile.kt

```diff
@@ -5,20 +5,54 @@ import com.intellij.openapi.fileTypes.FileType
 import com.intellij.openapi.util.io.FileUtil
 import com.intellij.openapi.vfs.VirtualFile
 import com.intellij.psi.FileViewProvider
+import com.intellij.psi.PsiDirectory
+import com.intellij.psi.PsiElement
 import com.intellij.psi.PsiFile
+import com.intellij.psi.util.PsiTreeUtil
+import org.rust.cargo.util.crateRoots
 import org.rust.lang.RustFileType
 import org.rust.lang.RustLanguage
-import org.rust.lang.core.psi.RustFileModItem
+import org.rust.lang.core.psi.RustDeclaringElement
+import org.rust.lang.core.psi.RustItem
+import org.rust.lang.core.psi.RustMod
+import org.rust.lang.core.psi.RustModDeclItem
 import org.rust.lang.core.psi.util.RustModules
+import org.rust.lang.core.psi.util.module
 import org.rust.lang.core.resolve.indexes.RustModulePath
+import org.rust.lang.core.resolve.indexes.RustModulesIndex
+import org.rust.lang.core.resolve.ref.RustReference
 
-class RustFile(fileViewProvider: FileViewProvider) : PsiFileBase(fileViewProvider, RustLanguage) {
+class RustFile(fileViewProvider: FileViewProvider) : PsiFileBase(fileViewProvider, RustLanguage), RustMod {
+    override fun getReference(): RustReference? = null
+
+    override val nameElement: PsiElement? = null
 
     override fun getFileType(): FileType = RustFileType
 
-    val mod: RustFileModItem?
-        get() = findChildByClass(RustFileModItem::class.java)
+    override val items: List<RustItem>
+        get() = PsiTreeUtil.getChildrenOfTypeAsList(this, RustItem::class.java)
```

Comment:
This can be extracted to a separate extension method which uses reified parameters, like `parentOfType`